h2o-core/src/main/java/water/rapids/ast/prims/advmath/SpearmanCorrelation.java

```diff
@@ -1,72 +1,132 @@
 package water.rapids.ast.prims.advmath;
 
+import water.Key;
 import water.MRTask;
 import water.Scope;
 import water.fvec.Chunk;
 import water.fvec.Frame;
 import water.fvec.Vec;
-import water.rapids.Env;
 import water.rapids.Merge;
-import water.rapids.Val;
-import water.rapids.ast.AstPrimitive;
-import water.rapids.ast.AstRoot;
-import water.rapids.vals.ValNum;
 import water.util.FrameUtils;
-import water.util.VecUtils;
 
 import java.math.BigDecimal;
 import java.math.MathContext;
+import java.util.Objects;
+
+public class SpearmanCorrelation {
+
+  public static Frame calculate(final Frame frameX, Frame frameY, final AstCorrelation.Mode mode) {
+    Objects.requireNonNull(frameX);
+    Objects.requireNonNull(frameY);
+
+    final Frame correlationMatrix = createCorrelationMatrix(frameX);
+    final boolean framesAreEqual = frameX == frameY; // TODO: Client APIs create new frames with no keys to compare
+
+    checkCorrelationDoable(frameX, frameY, mode);
+
+    for (int vecIdX = 0; vecIdX < frameX.numCols(); vecIdX++) {
+      for (int vecIdY = 0; vecIdY < frameX.numCols(); vecIdY++) {
```

Comment:
should there be frameY.numCols, I know they are same, in which case I would make that more explicit, or maybe this is fine, just caught my eye as a bug at first look
maybe checkCorrelationDoable() can return ncols and that can be used here?

Reply:
Thank you, it would be useful indeed to be more expressive about that.

I chose the way of doing `frameY.numCols()` in the inner iteration.

Also added `assert frameX.numCols() == frameY.numCols();` at the very beginning of the method after null checks to make it clear what is expected.